### ISSUE TITLE
Crash on 3d ionization

### ISSUE THREAD

**Issue by fyli16** [SmileiPIC/Smilei#45]
Hi, I have met a strange problem with 3d ionization.

Basically I was trying laser ionization of a spheric target, but SMILEI crashed as long as the sphere radius (or volume) was big enough. For instance, the attached test input (inp.txt) works fine for radius of 0.1 laser wavelength (line 39, if R<0.1), but crashes for R<0.2!

I have tested 2d ionization and 3d preionized plasma, and both worked fine.

I have also tested varying number of patches and MPI thread numbers for the crashed case, and they didn't help.

The log files for the crashed simulation are attached as out_*.  

The log file (make.txt) for compiling the code on my server (Ubuntu OS) is attached as well.

Could anyone have any ideas on this problem. Many thanks.

[inp.txt](https://github.com/SmileiPIC/Smilei/files/1845779/inp.txt)
[make.txt](https://github.com/SmileiPIC/Smilei/files/1845780/make.txt)
[out_e.txt](https://github.com/SmileiPIC/Smilei/files/1845781/out_e.txt)
[out_o.txt](https://github.com/SmileiPIC/Smilei/files/1845782/out_o.txt)

**Comment by fyli16** [comment]
BTW, I used 20 cores with 20 MPI processes in total for this small-scale test. It should not be a problem with the computation load, as larger scale tests with preionized plasma have worked perfectly.

Thanks.

**Comment by fyli16** [comment]
And I compiled the program with openmpi-1.10.2, hdf5-1.8.16 as indicated by the manual. Python was provided by Anaconda2.

**Comment by mccoys** [comment]
Hi, this looks like new ionized particles cause projection errors.

Have you tried to use the version tagged 3.3 ? You are likely using the latest version provided on the github repository, but it is not considered as stable.

It might also be an issue with the load balancing causing bad particle sorting. It could be interesting to remove load balancing to verify.

We will look into it as soon as possible

**Comment by mccoys** [comment]
I managed to run your attached case successfully on both 16 and 20 procs, several times in a row.

Does the error appear every time you run the code?

Which C++ compiler ? Which version of python exactly ?

**Comment by fyli16** [comment]
Hi mccoys, 
thanks a lot for the prompt response and help.

- yes, I used the latest github repository. Just now I compiled the v3.3 version from the 'release' page, and found the problem persisted with this version, generating the same errors. 

- The simulation generally always stopped at steps=87, corresponding to t=2.5 laser periods, for target radius of 0.2 laser wavelengths. It changed to steps=83 for radius of 0.3. But in either case, the laser beam has not even touched the target, so in principle, no ionization should have happened, and it might not be a projection issue with the new ionized particles.

- I tried by removing the load balancing module, and it didn't help.

- Thanks for testing the problem yourself. In my server, the problem is persistent; I tried many times, which all crashed for R>0.2.

- I use gcc 5.4.0 in my server. The c++11 compiler should be enabled with 'CXXFLAGS += -std=c++11' as given in the makefile, if I understand correctly.

- The system python2 version is 2.7.12. 'numpy' was enabled with Anaconda2, which has a python of version 2.7.13.

Thanks.

**Comment by mccoys** [comment]
Everything should be working with your setup. Have you tried running on more nodes, with more memory?

 Have you tried removing ionization ?

As I cannot reproduce the error, it is difficult to say, but this kind of error usually happens when a particle is not in the domain of the processor. I has not been correctly exchanged. I would suspect first the ionization, even though there should not be any ionization yet (but you never know). The fact that there is some randomness also tends to indicate ionization. Could you test without ionization?

**Comment by jderouillat** [comment]
I did not succeed to reproduce it with an Intel compiler but I succeed with g++. It's the same crash than yours.  
I'm working on it.

**Comment by fyli16** [comment]
@mccoys Yes, it worked fine without ionization. Running with more nodes/memory was not helping.

Great to see it might be related to the GNU compilers. I have intel compilers installed, could you pls indicate how I can modify the makefile to enable compilation with intel? Thanks for your time.

**Comment by mccoys** [comment]
@fyli16 To use the intel compiler, you simply need to make sure that the ``mpicxx`` executable is the intel version. Otherwise, you can define the environment variable ``SMILEICXX`` to the executable of your choice.

We will still continue to investigate the issue with gcc.

**Comment by iltommi** [comment]
this will tell openmpi to use the intel c++ compiler
```
export OMPI_CXX=icpc
```

**Comment by fyli16** [comment]
@iltommi @jderouillat @mccoys thank you all for the kind help. I have tried intel c++, and it is working now. Would like to hear about the solutions to gcc when you have. Thanks.

**Comment by jderouillat** [comment]
I find a bug in the 3D ionization projector, it has been fixed.

**Comment by mccoys** [comment]
It was only buggy for the gcc compiler ?!

**Comment by jderouillat** [comment]
No :  
```C++
-    zpn    = particles.position(1, ipart) * dz_inv_;  // normalized distance to the first node
+    zpn    = particles.position(2, ipart) * dz_inv_;  // normalized distance to the first node
```

**Comment by mccoys** [comment]
I know, but why would it work for intel then ???

**Comment by jderouillat** [comment]
It ran but we don't know if results were good ...

**Comment by mccoys** [comment]
@fyli16 have you been able to try again gcc with the corrected version ?

**Comment by fyli16** [comment]
@mccoys yes, I have tried gcc with the corrected version and it's working now. Thanks a lot.
